### ISSUE TITLE
Merging issue #9 and #10

### ISSUE THREAD

**Issue by Bistard** [mdbetancourt/eslint-plugin-neverthrow#15]
I can create a PR to help solve #9 and #10. I need a confirm that this repository is still worth updating. I hope anyone is still maintaining this.

**Comment by mdbetancourt** [comment]
Hi @Bistard thanks for you interest, i'm not mantaining this repository actively but if you create the PR solving those issue i'll merge it

**Comment by Bistard** [comment]
> Hi @Bistard thanks for you interest, i'm not mantaining this repository actively but if you create the PR solving those issue i'll merge it

I did a lot of changes to the implementation of `eslint-plugin-neverthrow`. Not just checking new corner cases but also making some different assumptions.

Here are some small changes I did:
1. I have my own `Result` type in my project, which has more methods that can be considered as handled: `match`, `unwrap`, `unwrapOr`, `unwrapErr`,`isOk`, `isErr`, `expect`,`then` (only avaliable for `AsyncResult`). This also helps to solve the issue #9.
2. Adding `AwaitExpression` parsing check that can easily solve #10 
3. If the unhandled `Result` is detected, the linter is not reporting to the original expression, it will try to report the first assigned variable.
```ts
{
        const firstRef = getResult(); // `firstRef` will be red-underlined, not `getResult()`
        const secondRef = firstRef;
}
```
4. This eslint rule also works with `AsyncResult` in all cases.

Here is one big assumption i made differently from this repo:
1. Quoting [this](https://github.com/supermacro/neverthrow/issues/212#issuecomment-953399008), **I am assuming passing `Result` to function is considered as taking ownership**. Making this assumption brings me to the following a more robust handling check.
2. Now linter will check if `Result` which in function parameter is handled, if not, error reported.
```ts
// this parameters will be reported as unhandled.

function resultInParameter(res2: Result<void, void>): void {}
    
async function asyncResultInParameter(res: AsyncResult<void, void>): Promise<void> {}

const anonymousFunction = function (res: Result<void, void>) {};

const anonymousArrowFunction = (res: Result<void, void>) => {};

class SomeClass {
    public classMethod(res: Result<void, void>) {}
}
```

@supermacro @mdbetancourt It depends on what kind of changes you guys want me to merge, either all changes, all making some of them as optional, or simply not accepting some changes.

I am ok with cases.

**Comment by thenbe** [comment]
@Bistard Would you be able to to open a PR for #10? Since it's a [simple](https://github.com/supermacro/neverthrow/issues/212#issuecomment-1870706050) change, perhaps we can get it merged quickly.

**Comment by Bistard** [comment]
> @Bistard Would you be able to to open a PR for #10? Since it's a [simple](https://github.com/supermacro/neverthrow/issues/212#issuecomment-1870706050) change, perhaps we can get it merged quickly.

Just did (here)[https://github.com/mdbetancourt/eslint-plugin-neverthrow/pull/17]

**Comment by Bistard** [comment]
Once the above PR is merged. I will continue to create a new PR that fixes #9. In additions, I need a confirm from the maintainer (@mdbetancourt, @supermacro) that should I perform the following big change:
> Assuming passing Result to function is considered as taking ownership.

If that is true, I will also complete the linting rule to check the function parameter Result handling.
